extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -237,6 +261,101 @@ export class AmpAdExit extends AMP.BaseElement {
       this.user().error(TAG, 'Invalid JSON config', e);
       throw e;
     }
+
+    this.ampAdResourceId_ = user().assert(
+        this.getAmpAdResourceId_(),
+        `${TAG}: No friendly parent amp-ad element was found for amp-ad-exit.`);
+
+    this.init3pResponseListener_();
+  }
+
+  /**
+   * Gets the resource ID of the amp-ad element containing this amp-ad-exit tag.
+   * This is a pass-through for the version in service.js, solely because
+   * the one in service.js isn't stubbable for testing, since only object
+   * methods are stubbable.
+   * @returns {?string}
+   * @private
+   * @VisibleForTesting
+   */
+  getAmpAdResourceId_() {
+    return getAmpAdResourceId(this.element, this.win.top);
+  }
+
+  /** @override */
+  resumeCallback() {
+    this.init3pResponseListener_();
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.unlisten_) {
+      this.unlisten_();
+      this.unlisten_ = null;
+    }
+    return super.unlayoutCallback();
+  }
+
+  /**
+   * amp-analytics will create an iframe for vendors in
+   * extensions/amp-analytics/0.1/vendors.js who have transport/iframe defined.
+   * This is limited to MRC-accreddited vendors. The frame is removed in
+   * amp-analytics, and the listener is destroyed here, if the user
+   * navigates/swipes away from the page. Both are recreated if the user
+   * navigates back to the page.
+   * @private
+   */
+  init3pResponseListener_() {
+    this.unlisten_ = this.unlisten_ || listen(this.getAmpDoc().win, 'message',
+        event => {
+          const responseMsg = deserializeMessage(getData(event));
```

Comment:
You need source origin checks.

Reply:
Lower down in this method we call assertValidResponseMessage_() which in turn calls assertOriginMatchesVendor_()

Reply:
That doesn't seem to be sufficient. The aim here is:

1. To avoid calling JSON.parse on someone else's data.
   - This is **very** expensive, and a **must** before this gets merged.
2. To validate the message came from the iframe.
   - The check in `#assertValidResponseMessage` isn't sufficient, since I can post a message that will pass from any iframe on the page.
   - The check must be explicitly for `event.origin === origin` (where `origin` is known before we create the iframe)

Reply:
1. I think it's pretty central that we accept multiple key/value pairs in the response message. I can double-check that. But if so, we need them to be in some wire format or another. Honestly I don't really care if that's { a: b, c: d} versus a=b&c=d versus something else, but surely the expense of parsing one format would pretty much equal the expense of parsing the other format, so I think format is moot. Let me know if I'm missing something there.
These are MRC-accredited vendors. We've made it super clear that if they bog us down such that we can't maintain 60fps, we'll stop serving their iframe. So it's in their interest to behave. And there will likely be far fewer than 10 of them, ever (insert 640KB joke here). I suppose if you like, we can put a character limit on the response, and only attempt to JSON.parse the first max(length, 2083) characters or something? (The point of the response values are to build a URL, and 2083 = maximum URL length supported by IE10)

2. We are doing what you are saying here...but perhaps I did not explain it well. Each 3p vendor's (non-friendly) iframe lives at a certain URL. This URL must be specified in extensions/amp-analytics/0.1/vendors.js, as follows:
```
  'some-vendor': {
    'transport': {
      'iframe': 'https://somevendor.com/page.html',
    },
  },
```
This is subject to the ampproject approval process, and all that. AmpAnalytics looks there, to know what the src should be when it creates the iframe. So, that's the ground truth for the "origin" in the right-hand side of your event.origin === origin.
To trace the code for the left-hand-side of this equality:
```
this.assertValidResponseMessage_(responseMsg, event.origin);
...
// In assertValidResponseMessage_, 2nd param gets called "responseOrigin":
this.assertOriginMatchesVendor_(responseOrigin, responseMessage['vendor']);
...
// In assertOriginMatchesVendor_, 1st param gets called "origin":
    user().assert(vendorURL && vendorURL.origin == origin,
```
...so I hope I've shown we're comparing *something* to event.origin.

To show that we're comparing event.origin to the ground-truth origin of where the frame was created (the right-hand-side of your equality):
vendorUrl in the previous snippet comes from:
```
    const vendorURL = parseUrl(assertVendor(vendor));
```
Assume for the moment that vendor is correct.
assertVendor() simply pulls the ground-truth iframe URL associated with that vendor from vendors.js, asserts that it exists, and returns it:
```
  user().assert(ANALYTICS_CONFIG &&
      ANALYTICS_CONFIG[vendor] &&
      ANALYTICS_CONFIG[vendor]['transport'] &&
      ANALYTICS_CONFIG[vendor]['transport']['iframe'],
      'Unknown vendor: ' + vendor);
  return ANALYTICS_CONFIG[vendor]['transport']['iframe'];
```
So (if "vendor" is to be believed) we have compared the event.origin to what we know the origin should be for that vendor, which was known prior to creating the iframe.
I posit that "vendor" should be believed, as it comes from the "name" attribute that amp-analytics set on the iframe when it created it (see https://github.com/ampproject/amphtml/blob/master/3p/iframe-transport-client.js#L42 and line 164 of same file).

Reply:
I spoke with Keith, and he interpreted your first point differently than I did. Are you just saying don't call deserializeMessage() (which calls JSON.parse()) until *after* the origin check? I had thought you meant we can't call it at all, which would kind of torpedo the whole project.

Reply:
Thinking about this a little more...to do the origin check, we'd need to know the vendor, and we don't get the vendor until after we parse the JSON. But I can see a couple of workarounds:
1. We can see event.origin without deserializing/parsing event.data. We could at least verify that event.origin is a valid 3p vendor origin for some vendor. Or better yet, for some vendor that is referenced in this amp-ad-exit's config.
2. This is kludgey, but if that's not sufficient, we could use a simple regex to try to get the vendor out of event.data without parsing the whole thing. An example of event.data is:
```
"amp-1507302338913{"message":{"collected-data":"abc"},"creativeId":"1","vendor":"example-3p-vendor","type":"iframe-transport-response","sentinel":"1"}"
```
We could check it with something like /"vendor":"([^"+])/ and then ensure that the matched text is a vendor name.

Reply:
Ok, I changed things such that now we build expectedOriginToVendor_ when parsing the amp-ad-exit config, such that it contains origin->vendor for the vendors that appear in the config.
When we receive an event, we check that event.origin is in there before parsing the event data.